src/shared/sleep-config.c

```diff
@@ -65,64 +78,26 @@ int parse_sleep_config(const char *verb, bool *ret_allow, char ***ret_modes, cha
                                         "Sleep\0", config_item_table_lookup, items,
                                         CONFIG_PARSE_WARN, NULL);
 
-        if (streq(verb, "suspend")) {
-                allow = allow_suspend != 0;
-
-                /* empty by default */
-                modes = TAKE_PTR(suspend_mode);
-
-                if (suspend_state)
-                        states = TAKE_PTR(suspend_state);
-                else
-                        states = strv_new("mem", "standby", "freeze");
-
-        } else if (streq(verb, "hibernate")) {
-                allow = allow_hibernate != 0;
-
-                if (hibernate_mode)
-                        modes = TAKE_PTR(hibernate_mode);
-                else
-                        modes = strv_new("platform", "shutdown");
-
-                if (hibernate_state)
-                        states = TAKE_PTR(hibernate_state);
-                else
-                        states = strv_new("disk");
-
-        } else if (streq(verb, "hybrid-sleep")) {
-                allow = allow_hybrid_sleep > 0 ||
-                        (allow_suspend != 0 && allow_hibernate != 0);
-
-                if (hybrid_mode)
-                        modes = TAKE_PTR(hybrid_mode);
-                else
-                        modes = strv_new("suspend", "platform", "shutdown");
-
-                if (hybrid_state)
-                        states = TAKE_PTR(hybrid_state);
-                else
-                        states = strv_new("disk");
-
-        } else if (streq(verb, "suspend-then-hibernate")) {
-                allow = allow_s2h > 0 ||
-                        (allow_suspend != 0 && allow_hibernate != 0);
```

Comment:
So... the code before calculated the `allow` value based on the verb. You dropped this logic completely. For example, in the case of s2h, `allow` is true if allow_s2h is explicitly set in the config file OR neither suspend nor hibernation is disallowed. But after your patch, it is also implicitly allowed (equal to -1).

(Actually, it could be said the existing code was buggy. It should have been something like
```C
allow = allow_s2h >= 0 ? allow_s2h : allow_suspend != 0 && allow_hibernate != 0;
```
i.e. an explicit setting should always win. This should be fixed in a separate commit.)

Reply:
Good catch! Thanks.

Reply:
Good catch! I moved this logic to the new `sleep_settings` function but your comment makes me realize that the `SleepConfig` instance can have unwanted implicit state if the caller just opts to check `sleep_config->allow_` rather than calling `sleep_settings` to set  `ret_allow`. I'll move the logic back to the `parse_` function so that the `SleepConfig` instance has a valid state.

Reply:
@keszybz, maybe I'm reading the original code wrong, but it looks like "implicit allow" is the intended default behaviour for most of the `allow` checks. The `allow_` values are initialized to `-1` and then assigned via `config_parse_tristate` which allows for `-1` (`uninitialized`).  The logic then mostly checks for `allow_X != 0` which implies that `-1` would default to `true` in those cases. The exceptions are `allow_hybrid_sleep` and `allow_s2h` which check for `> 0` which implies that the default value is `false` in those cases.

If we want the caller to be able to check `sleep_config->allow_` outside of calling `sleep_settings` I should refactor the `SleepConfig` type to use `bool` instead of `int` for the `allow_` fields - otherwise, `-1` would be a valid value and the caller would be required to understand when `-1` is `true` vs. `false` which seems like a recipe for bugs.

I'll move the `allow` logic back into `parse_sleep_config` from `sleep_settings` and refactor the `SleepConfig` type to use `bool` over `int` so that `sleep_config->allow_` holds the correct value for each case.

As for fixing the `allow_s2h` logic, that make sense to me. Just so I'm understanding correctly, are you suggesting that change be in a single non-squashed commit so it is clear that the logic was intentionally changed to fix a bug in the commit history?

Reply:
>  The logic then mostly checks for allow_X != 0 which implies that -1 would default to true in those cases.

Yes.

> The exceptions are allow_hybrid_sleep and allow_s2h which check for > 0 which implies that the default value is false in those cases.

Not really. The second clause in the conditional usually wins (`allow_suspend != 0 && allow_hibernate != 0`), which means that unless the user explicitly disabled suspend or hibernation, they'll get s2h too.

> If we want the caller to be able to check sleep_config->allow_ outside of calling sleep_settings I should refactor the SleepConfig type to use bool instead of int for the allow_ fields - otherwise, -1 would be a valid value and the caller would be required to understand when -1 is true vs. false which seems like a recipe for bugs.

Sure. It makes sense to only store booleans in the structure.

> Just so I'm understanding correctly, are you suggesting that change be in a single non-squashed commit so it is clear that the logic was intentionally changed to fix a bug in the commit history?

Exactly.

Reply:
Sounds good. I'll work through these changes soon. Thanks!